src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -915,21 +906,137 @@ public override void ReportDiagnosticsIfObsolete(Binder binder, SyntaxNode synta
 
             protected override bool TypeSymbolEquals(TypeSymbolWithAnnotations other, TypeCompareKind comparison)
             {
-                var otherLazy = other as LazyNullableType;
-
+                var otherLazy = other as LazyNullableTypeParameter;
                 if ((object)otherLazy != null)
                 {
                     return _underlying.TypeSymbolEquals(otherLazy._underlying, comparison);
                 }
+                return base.TypeSymbolEquals(other, comparison);
+            }
+        }
+
+        /// <summary>
+        /// Non-nullable or null-oblivious type parameter. IsNullable is calculated
+        /// lazily to avoid cycles when binding declarations.
+        /// </summary>
+        private sealed class LazyNonNullableTypeParameter : TypeSymbolWithAnnotations
+        {
+            private readonly TypeParameterSymbol _typeParameter;
+            private readonly bool? _isNullableIfReferenceType;
+
+            internal LazyNonNullableTypeParameter(TypeParameterSymbol typeParameter, bool? isNullableIfReferenceType)
+            {
+                Debug.Assert(isNullableIfReferenceType != true); // Should use LazyNullableTypeParameter.
```

Comment:
PROTOTYPE on the comment?

Reply:
The comment was intended to indicate why the assert failed rather than work that remained. I've updated the comment to clarify.

---
In reply to: [201549604](https://github.com/dotnet/roslyn/pull/28453#discussion_r201549604) [](ancestors = 201549604)